OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs

```diff
@@ -70,16 +93,21 @@ void INotifyDamage.Damaged(Actor self, AttackInfo e)
 			if (!info.DamageTypes.IsEmpty && !e.Damage.DamageTypes.Overlaps(info.DamageTypes))
 				return;
 
-			if (isSmoking) return;
+			if (isPlayingAnimation) return;
 			if (e.Damage.Value < 0) return; /* getting healed */
 			if (e.DamageState < info.MinimumDamageState) return;
 			if (e.DamageState > info.MaximumDamageState) return;
 
-			isSmoking = true;
-			anim.PlayThen(info.IdleSequence,
-				() => anim.PlayThen(info.LoopSequence,
-					() => anim.PlayThen(info.EndSequence,
-						() => isSmoking = false)));
+			isPlayingAnimation = true;
+			PlayAnimation(info.GetSequences(self.World.LocalRandom).GetEnumerator());
+		}
+
+		void PlayAnimation(IEnumerator<string> sequences)
+		{
+			if (sequences.MoveNext())
+				anim.PlayThen(sequences.Current, () => PlayAnimation(sequences));
```

Comment:
Recursively calling the same function with the same parameters. Won't cause issues with memory retention, or possibly endless loops? Feels bad. Even with a few sequences would waste some stack frames.

Reply:
It's just a deconstructed foreach loop